### PR TITLE
Silence all warnings when building with GCC on Linux

### PR DESCRIPTION
**Pull request by giordano**
With this PR I can do a clean build from scratch (at least after `git clean -fxd`) with `make CFLAGS=-Werror CXXFLAGS=-Werror` when using GCC 11.1.0 on Linux.  Clang is _way_ more picky.

@staticfloat @vtjnash does it make sense to enable `-Werror` in CI at least for the platforms where we use GCC?  Now that there is a (partially) clean state it'd be good to not introduce new warnings.

### CODE REVIEW

**Comment by giordano** [comment]
In https://build.julialang.org/#/builders/41/builds/3390/steps/7/logs/stdio I see another warning which I didn't get locally
```
In file included from /buildworker/worker/package_linux64/build/src/codegen.cpp:1673:0:
/buildworker/worker/package_linux64/build/src/cgutils.cpp: In function 'jl_cgval_t emit_setfield(jl_codectx_t&, jl_datatype_t*, const jl_cgval_t&, size_t, jl_cgval_t, jl_cgval_t, bool, llvm::AtomicOrdering, llvm::AtomicOrdering, bool, bool, bool, bool, bool, const jl_cgval_t*, const string&)':
/buildworker/worker/package_linux64/build/src/cgutils.cpp:3339:37: warning: 'ModifyBB' may be used uninitialized in this function [-Wmaybe-uninitialized]
             ctx.builder.CreateCondBr(Success, XchgBB, ismodifyfield ? ModifyBB : DoneBB);
             ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
`ModifyBB` is set at
https://github.com/JuliaLang/julia/blob/a2e4226d2e47a38ca4f15cfc69504b6e16631d79/src/cgutils.cpp#L3302-L3306 and then used at https://github.com/JuliaLang/julia/blob/a2e4226d2e47a38ca4f15cfc69504b6e16631d79/src/cgutils.cpp#L3339
 only if `ismodifyfield` is true, which is the same condition when it's set.  I wonder if newer GCC got smarter?

**Comment by giordano** [comment]
Instead [on aarch64](https://build.julialang.org/#/builders/11/builds/3590/steps/7/logs/stdio) we get plenty of warnings like
```
In file included from /buildworker/worker/package_linuxaarch64/build/usr/include/libunwind.h:7:0,
                 from /buildworker/worker/package_linuxaarch64/build/src/julia_internal.h:1019,
                 from /buildworker/worker/package_linuxaarch64/build/src/jloptions.c:7:
/buildworker/worker/package_linuxaarch64/build/usr/include/libunwind-aarch64.h:60:9: warning: empty struct has size 0 in C, size 1 in C++ [-Wc++-compat]
 typedef struct
         ^~~~~~
/buildworker/worker/package_linuxaarch64/build/usr/include/libunwind-aarch64.h:169:16: warning: empty struct has size 0 in C, size 1 in C++ [-Wc++-compat]
 typedef struct unw_tdep_save_loc
                ^~~~~~~~~~~~~~~~~
```
which come from libunwind...  I guess at the moment we can do little about them apart from disabling the corresponding warnings

**Comment by vtjnash** [comment]
libunwind is usually pretty good about taking patches. Looks like some platforms put `char unused;` there, but not aarch64 or x86 and probably others.

**Comment by DilumAluthge** [comment]
> Again, what about enabling `-Werror` in CI at least for linux64 now that it's clean?

This makes sense to me! And then we can add other platforms one by one as we fix then.

@staticfloat @vtjnash Are you cool with this idea? If so, I'll give @giordano directions to the two places that he'll need to implement this change.

**Comment by giordano** [comment]
Bump :slightly_smiling_face:

**Comment by staticfloat** [comment]
Yeah, I'm cool with this.  I say merge this now, and then add `-Werror` after we transition to buildkite within the next day or two.

**Comment by giordano** [comment]
Does anyone want to review the changes?  I think they should all be mostly uncontroversial, the only one I'm not entire sure is `#warning` -> `#pragma message`, in case there are other better options.

File: src/gc-debug.c [hunk 1]
```diff
@@ -983,13 +983,13 @@ void gc_time_summary(int sweep_full, uint64_t start, uint64_t end,
...
-                       live = %ldm new interval = %ldm time = %ldms\n",
+        jl_safe_printf("%ld Major collection: estimate freed = %ld\n"
+                       "live = %ldm new interval = %ldm time = %ldms\n",
                        end - start, freed, live/1024/1024,
                        interval/1024/1024, pause/1000000 );
     else
-        jl_safe_printf("%ld Minor collection: estimate freed = %ld live = %ldm
-                       new interval = %ldm time = %ldms\n",
+        jl_safe_printf("%ld Minor collection: estimate freed = %ld live = %ldm\n"
+                       "new interval = %ldm time = %ldms\n",
```
Comment (giordano) [inline src/gc-debug.c:992]: I'm not even sure how this compiled, but I presume this was the intention of these printfs?

**Comment by imciner2** [comment]
> Yeah, I'm cool with this. I say merge this now, and then add -Werror after we transition to buildkite within the next day or two.

Can this be made conditional for use on the CI systems only? Adding `-Werror` makes packaging in linux distros that either inject their own compiler warnings or update compilers more often more complicated because usually with a new compiler comes a new warning or two that are enabled by default (or a change in the way warning semantics are used).

**Comment by giordano** [comment]
> Can this be made conditional for use on the CI systems only?

I don't think anyone suggested to make `-Werror` default for everybody, which I agree would be disruptive.  The proposal here is to only use it in CI.

**Comment by imciner2** [comment]
> I don't think anyone suggested to make -Werror default for everybody, which I agree would be disruptive. The proposal here is to only use it in CI.

Ok, that's good. I had started reading from where GitHub told me the new messages were and had forgotten the prior mentions of it in the thread since it had been so long :laughing:.

**Comment by giordano** [comment]
Besides the warnings fixed by #44669 (introduced #44228), there are now new warnings introduced by #44634.  This is a whack-a-mole game

File: src/gc-debug.c (hunk 1 above)
Comment (vtjnash) [reply src/gc-debug.c:992]: I am not sure it intended to have newlines

Comment (giordano) [reply src/gc-debug.c:992]: @chflood how where these lines intended to be printed?  A standard `printf` wouldn't even compile with the previous syntax.  I'm happy to fix it according to the original intention.

**Comment by chflood** [comment]
The problem with the original code was that it didn't print the right
numbers on 32 bit Julia.  The problem with the current code is that it gets
a warning on 64 bit Julia.  I will take a look.

**Pull request merged**